{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/thihanmoekyaw/NCSA_Manhattan_Plot/blob/master/example/demo.ipynb)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !pip install dash==2.8.1\n",
    "# !pip install jupyter_dash==0.4.2\n",
    "# !pip install plotly.express==0.4.1\n",
    "# !pip install numpy==1.22.4\n",
    "# !pip install pandas==1.3.5\n",
    "!pip install git+https://github.com/thihanmoekyaw/NCSA_Manhattan_Plot.git"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import plotly.graph_objs as go\n",
    "import re\n",
    "from dash.dependencies import Input, Output, State\n",
    "from dash import html, dcc, dash_table\n",
    "from jupyter_dash import JupyterDash\n",
    "# This package\n",
    "from NCSA_Manhattan_Plot import ManhattanPlot as NMP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "neg_log_10_p_val    float64\n",
      "var                  object\n",
      "CHR                   int64\n",
      "Position              int64\n",
      "Gene                 object\n",
      "alias                object\n",
      "Kegg_Pathway         object\n",
      "dtype: object\n",
      "Note: this may take a while.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"100%\"\n",
       "            height=\"650\"\n",
       "            src=\"http://127.0.0.1:8050/\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "            \n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x17348d2a0>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import ssl\n",
    "\n",
    "ssl._create_default_https_context = ssl._create_unverified_context\n",
    "\n",
    "# start = time.process_time()\n",
    "url = \"https://raw.githubusercontent.com/thihanmoekyaw/NCSA_Manhattan_Plot/main/example/demo-gwas-data.csv\"\n",
    "df = pd.read_csv(url)\n",
    "print(df.dtypes)\n",
    "bonferroni_val = 5\n",
    "Benjamini_Hochberg_val = 7\n",
    "\n",
    "# Change the data type of the specified columns from object to string\n",
    "df['var'] = df['var'].astype(str)\n",
    "df['Gene'] = df['Gene'].astype(str)\n",
    "df['alias'] = df['alias'].astype(str)\n",
    "df['Kegg_Pathway'] = df['Kegg_Pathway'].astype(str)\n",
    "\n",
    "# Get all unique pathway names by splitting the Kegg_Pathway column\n",
    "unique_pathways = set()\n",
    "for pname in df['Kegg_Pathway']:\n",
    "    pathways = re.split(r';+', pname)\n",
    "    unique_pathways.update(pathways)\n",
    "# print(unique_pathways)\n",
    "\n",
    "MahanttanPlot = NMP.NCSA_ManhattanPlot(\n",
    "    dataframe=df, \n",
    "    chrm = \"CHR\", \n",
    "    bp = \"Position\", \n",
    "    p = \"neg_log_10_p_val\", \n",
    "    snp = \"var\",\n",
    "    gene = \"alias\", \n",
    "    logp = False, \n",
    "    annotation=\"Kegg_Pathway\",\n",
    "    genomewideline_value = 5,\n",
    "    suggestiveline_value = 7,\n",
    "    additional_highlight_row=\"Kegg_Pathway\", #the name of the row in pandas\n",
    "    additional_highlight_val=\"RNA transport\", #either a string or a list\n",
    ")\n",
    "\n",
    "MahanttanPlot.update_layout(\n",
    "    # width=1000,  # set the width of the graph\n",
    "    # height=700  # set the height of the graph\n",
    ")\n",
    "\n",
    "Mahanttan_app = JupyterDash(__name__)\n",
    "\n",
    "# add html components and figure to app\n",
    "Mahanttan_app.layout = html.Div([\n",
    "    # Add the dropdown menu for PNAME\n",
    "    html.Div([\n",
    "        html.Div(\"For the threshold line, what value would you like to use?\", style={'display': 'inline-block', 'paddingRight': '10px'}),\n",
    "        dcc.RadioItems(\n",
    "            id='threshold-radio',\n",
    "            options=[\n",
    "                {'label': 'Slider value', 'value': 'slider'},\n",
    "                {'label': 'Bonferroni correction', 'value': 7}\n",
    "            ],\n",
    "            value='slider',\n",
    "            style={\n",
    "                'display': 'block',\n",
    "                'verticalAlign': 'middle'\n",
    "            }\n",
    "        ),\n",
    "    ]),\n",
    "    html.Br(),\n",
    "    dcc.Dropdown(\n",
    "        id='pname-dropdown',\n",
    "        options=[{'label': pname, 'value': pname} for pname in unique_pathways],\n",
    "        value=list(unique_pathways)[0],\n",
    "        multi=True\n",
    "    ),\n",
    "    html.Br(),\n",
    "    dcc.Slider(\n",
    "        id='threshold-slider',\n",
    "        min=1,\n",
    "        max=10,\n",
    "        marks={\n",
    "            i: {'label': str(i)} for i in range(10)\n",
    "        },\n",
    "        value=5\n",
    "    ),\n",
    "    html.Br(),\n",
    "    dcc.Graph(id='mahanttan-graph', figure=MahanttanPlot),\n",
    "    html.Div('''\n",
    "    note: due to how the Plotly Bio Package works, chromosome X, Y and MT where changed to chromosome 23, 24, 25 respectively.\n",
    "    The blue line is the Benjamini Hochberg critical p-value and the red line is the Bonferroni value.'''),\n",
    "    html.Div(id='mahanttan-output'),\n",
    "\n",
    "])\n",
    "\n",
    "# Callback function to update the Manhattan plot when the slider value changes\n",
    "@Mahanttan_app.callback(\n",
    "    Output('mahanttan-graph', 'figure'),\n",
    "    Input('pname-dropdown', 'value'),\n",
    "    Input('threshold-slider', 'value'),\n",
    "    Input('threshold-radio', 'value')\n",
    ")\n",
    "def update_mahanttan_plot(pname_value, threshold_value, threshold_radio_value):\n",
    "    if threshold_radio_value == 'slider':\n",
    "        genomewideline_val = threshold_value\n",
    "    elif threshold_radio_value == 'bonferroni':\n",
    "        genomewideline_val = bonferroni_val\n",
    "        \n",
    "    MahanttanPlot = NMP.NCSA_ManhattanPlot(\n",
    "        dataframe=df, \n",
    "        chrm = \"CHR\", \n",
    "        bp = \"Position\", \n",
    "        p = \"neg_log_10_p_val\", \n",
    "        snp = \"var\",\n",
    "        gene = \"alias\", \n",
    "        logp = False, \n",
    "        annotation=\"Kegg_Pathway\",\n",
    "        genomewideline_value=genomewideline_val,\n",
    "        suggestiveline_value=Benjamini_Hochberg_val,\n",
    "        additional_highlight_row=\"Kegg_Pathway\", #the name of the row in pandas\n",
    "        additional_highlight_val=pname_value, #either a string or a list\n",
    "        #in_pathway=\"IN_PATHWAY\"\n",
    "    )\n",
    "    MahanttanPlot.update_layout(\n",
    "    # width=1000,  # set the width of the graph\n",
    "    # height=700  # set the height of the graph\n",
    "    )  \n",
    "\n",
    "    MahanttanPlot.update_layout()\n",
    "\n",
    "    return MahanttanPlot\n",
    "\n",
    "print(\"Note: this may take a while.\")\n",
    "Mahanttan_app.run_server(mode='inline') \n",
    "\n",
    "#what if we merge the annotation and annotation_row thingy together so that it is easier to debug and everythig? It makes more sense this way. \n",
    "# Pip install link using github with NCSA mahantann plot, after you finish with the annotation and the value merging"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
